sklearn/decomposition/pca.py

```diff
@@ -260,52 +321,71 @@ def fit_transform(self, X, y=None):
         return U
 
     def _fit(self, X):
-        """Fit the model on X
+        """Dispatch to the right submethod depending on the chosen solver."""
+        X = check_array(X, dtype=[np.float64], ensure_2d=True,
+                        copy=self.copy)
 
-        Parameters
-        ----------
-        X: array-like, shape (n_samples, n_features)
-            Training vector, where n_samples in the number of samples and
-            n_features is the number of features.
+        # Handle n_components==None
+        if self.n_components is None:
+            n_components = X.shape[1]
+        else:
+            n_components = self.n_components
 
-        Returns
-        -------
-        U, s, V : ndarrays
-            The SVD of the input data, copied and centered when
-            requested.
-        """
-        X = check_array(X)
+        # Handle svd_solver
+        svd_solver = self.svd_solver
+        if svd_solver == 'auto':
+            # Small problem, just call full PCA
+            if max(X.shape) <= 500:
```

Comment:
It is just a simple rule of thumb. PCA is anyway sufficiently fast on any modern machine when run on data of that size. Let's not forget that we are not only optimizing for speed, but also would like to keep using 'full 'PCA --which gives the exact answer- when possible.